src/main/java/explorableviz/transparenttext/AuthoringAssistant.java

```diff
@@ -25,27 +27,24 @@ public AuthoringAssistant(InContextLearning inContextLearning, String agentClass
         this.templateProgram = templateProgram;
     }
 
-    public List<Pair<Program, ProgramResult>> executePrograms() throws Exception {
-        List<Pair<Program, ProgramResult>> results = new ArrayList<>();
-        List<Pair<Program, Expression>> subTests = templateProgram.programs(templateProgram);
+    public List<Pair<Program, QueryResult>> executePrograms() throws Exception {
+        List<Pair<Program, QueryResult>> results = new ArrayList<>();
+        List<Pair<Program, Expression>> programEdits = templateProgram.asIndividualEdits(templateProgram);
         int i = 0;
-        while (!subTests.isEmpty()) {
-            Pair<Program, Expression> subTest = subTests.get(i);
-            Program subProgram = subTest.component1();
-            ProgramResult result = execute(subTest);
+        while (!programEdits.isEmpty()) {
+            Pair<Program, Expression> individualEdit = programEdits.get(i);
+            //selection
+            Program programEdit = individualEdit.component1();
+            QueryResult result = execute(individualEdit);
 
-            subProgram.replaceParagraph(subProgram.getParagraph().splice(result.response() == null ? subTest.component2() : result.response()));
-            results.add(new Pair<>(subProgram, result));
-            if(Settings.isEditorLoopEnabled()) {
-                subTests.addAll(subProgram.programs(templateProgram));
-            } else {
-                subTests = subProgram.programs(templateProgram);
-            }
+            programEdit.replaceParagraph(programEdit.getParagraph().splice(result.response() == null ? individualEdit.component2() : result.response()));
+            results.add(new Pair<>(programEdit, result));
+            programEdits = programEdit.asIndividualEdits(templateProgram);
         }
         return results;
     }
 
-    public ProgramResult execute(Pair<Program, Expression> test) throws Exception {
+    public QueryResult execute(Pair<Program, Expression> test) throws Exception {
```

Comment:
I think you forgot to remove the cast on line 59!

Reply:
Yeah I needed to adapt the call. Done